### PR TITLE
Specialize address space during spirv legalization.

### LINKED ISSUE #4380: SPIRV - OpAccessChain result type (OpTypePointer) does not match the type that results from indexing into the base <id> (OpTypePointer)
Hey, when I store a pointer inside a structure, I get errors from `spirv-val`.
```Slang
struct Constants {
  uint *p;
};

[[vk::push_constant]] Constants constants;

[[vk::binding(0)]] RWTexture2D<uint> my_texture;

struct MyType {
  uint *_p;

  __init(uint* p) {
    _p = p;
  }

  void write_to(RWTexture2D<uint> t) {
    t[uint2(0, 0)] = *_p;
  }
}

[numthreads(1, 1, 1)]
void main() {
  MyType(constants.p).write_to(my_texture);
}
```

compiling it with `slangc.exe -emit-spirv-directly .\bruh.slang -o bruh.spv | spirv-val.exe .\bruh.spv` returns errors from `spirv-val`:
```
error: line 49: OpVariable 60: expected AliasedPointer or RestrictPointer for PhysicalStorageBuffer pointer.
  %60 = OpVariable %_ptr_Function__ptr_PhysicalStorageBuffer_uint Function
```

The `slangc` version is `v2024.1.21` and `spirv-val` version is `SPIRV-Tools v2024.2 v2024.2.rc1-0-gdd4b663e`.

### LINKED ISSUE #4558: Modifying global static variable through `inout` parameter results invalid SPIRV.
Reproducer:
```
struct TT
{
    int a;
};

void modify(inout TT p)
{
    p.a = 1;
}

static TT r;
RWStructuredBuffer<int> output;

[numthreads(1, 1, 1)]
void main()
{
    modify(r);
    output[0] = r.a;
}
```

```
slangc.exe bug.hlsl -emit-spirv-directly -fvk-use-entrypoint-name -o out.spv -O0
PS D:\test> spirv-val out.spv
error: line 40: OpFunctionCall Argument <id> '20[%r]'s type does not match Function <id> '9[%_ptr_Function_TT]'s parameter type.
  %5 = OpFunctionCall %void %modify %r

```

### PR DESCRIPTION
**Pull request by csyonghe**
Closes #4558.
Closes #4380.

This change uses the address space specialization pass to propagate address space to all pointer values, and specialize the callees with address spaces from the argument.

### CODE REVIEW

File: docs/building.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ See the [documentation on testing](../tools/slang-test/README.md) for more infor
 | `SLANG_EMBED_STDLIB_SOURCE`       | `TRUE`           | Embed stdlib source in the binary                                  |
 | `SLANG_ENABLE_ASAN`               | `FALSE`          | Enable ASAN (address sanitizer)                                    |
 | `SLANG_ENABLE_FULL_IR_VALIDATION` | `FALSE`          | Enable full IR validation (SLOW!)                                  |
+| `SLANG_ENABLE_IR_BREAK_ALLOC`     | `FALSE`          | SLANG_ENABLE_IR_BREAK_ALLOC                                        |
```
Comment (jkwak-work) [inline docs/building.md:67]: I am not sure why this change is a part of this PR.
But the 3rd column should have an explanation of what the macro is rather than just repeating the macro keyword

**Review by jkwak-work — APPROVED** [review]
Looks good to me

File: source/slang/slang-ir.h [hunk 2]
```diff
@@ -41,7 +41,7 @@ struct  IRStructKey;
 
 enum class AddressSpace
 {
-    Generic = 0,
+    Generic = 0x7fffffff,
```
[the remark this answers is not in the document]
Comment (csyonghe) [reply source/slang/slang-ir.h:44]: This is just so that it is consistent with `SpvStorageClassMax`.

File: docs/building.md (hunk 1 above)
Comment (csyonghe) [reply docs/building.md]: Done.

Comment (csyonghe) [reply docs/building.md]: I needed this for debugging, somehow this is no longer enabled by default when in debug build after we switched to cmake.
